Assets/MixedRealityToolkit/Providers/BaseSpatialMeshObserver.cs

```diff
@@ -0,0 +1,279 @@
+            };
+
+        /// <summary>
+        /// Event sent whenever a mesh is discarded.
+        /// </summary>
+        protected static readonly ExecuteEvents.EventFunction<IMixedRealitySpatialAwarenessObservationHandler<SpatialAwarenessMeshObject>> OnMeshRemoved =
+            delegate (IMixedRealitySpatialAwarenessObservationHandler<SpatialAwarenessMeshObject> handler, BaseEventData eventData)
+            {
+                MixedRealitySpatialAwarenessEventData<SpatialAwarenessMeshObject> spatialEventData = ExecuteEvents.ValidateEventData<MixedRealitySpatialAwarenessEventData<SpatialAwarenessMeshObject>>(eventData);
+                handler.OnObservationRemoved(spatialEventData);
+            };
+
+        #endregion BaseSpatialMeshObserver Implementation
+
+        #region IMixedRealityDataProvider Implementation
+
+        /// <summary>
+        /// Initializes event data and creates the observer.
+        /// </summary>
+        public override void Initialize()
+        {
+            meshEventData = new MixedRealitySpatialAwarenessEventData<SpatialAwarenessMeshObject>(EventSystem.current);
+            
+            base.Initialize();
+        }
+
+        #endregion IMixedRealityDataProvider Implementation
+
+        #region IMixedRealitySpatialMeshObserver Implementation
+
+        private SpatialAwarenessMeshDisplayOptions displayOption = SpatialAwarenessMeshDisplayOptions.Visible;
+
+        /// <inheritdoc />
+        public SpatialAwarenessMeshDisplayOptions DisplayOption
+        {
+            get { return displayOption; }
+            set
+            {
+                displayOption = value;
+                ApplyUpdatedMeshDisplayOption(displayOption);
+            }
+        }
+
+        private SpatialAwarenessMeshLevelOfDetail levelOfDetail = SpatialAwarenessMeshLevelOfDetail.Coarse;
+
+        /// <inheritdoc />
+        public SpatialAwarenessMeshLevelOfDetail LevelOfDetail
+        {
+            get { return levelOfDetail; }
+            set
+            {
+                if (value != SpatialAwarenessMeshLevelOfDetail.Custom)
+                {
+                    TrianglesPerCubicMeter = LookupTriangleDensity(value);
+                }
+
+                levelOfDetail = value;
+            }
+        }
+
+        /// <summary>
+        /// The backing field for Meshes, to allow the mesh observer implementation to track its meshes.
+        /// </summary>
+        protected readonly Dictionary<int, SpatialAwarenessMeshObject> meshes = new Dictionary<int, SpatialAwarenessMeshObject>();
+
+        /// <inheritdoc />
+        public IReadOnlyDictionary<int, SpatialAwarenessMeshObject> Meshes => new Dictionary<int, SpatialAwarenessMeshObject>(meshes);
+
+        private int meshPhysicsLayer = 31;
+
+        /// <inheritdoc />
+        public int MeshPhysicsLayer
+        {
+            get { return meshPhysicsLayer; }
+            set
+            {
+                if ((value < 0) || (value > 31))
```

Comment:
I think I had 31 spatial awareness layer saved somewhere? use instead of hardcoded int

Reply:
and use in log error